include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,706 @@
+
+      AssertTS(TSSetSolution(ts, solution.petsc_vector()));
+
+      if (explicit_function)
+        AssertTS(
+          TSSetRHSFunction(ts, nullptr, _dealii_ts_rhsfunction_, nullptr));
+
+      if (implicit_function)
+        AssertTS(TSSetIFunction(ts, nullptr, _dealii_ts_ifunction_, nullptr));
+
+      if (setup_jacobian)
+        {
+          AssertTS(TSSetIJacobian(ts,
+                                  A ? A->petsc_matrix() : nullptr,
+                                  P ? P->petsc_matrix() : nullptr,
+                                  _dealii_ts_ijacobian_with_setup_,
+                                  nullptr));
+          SNES snes;
+          AssertTS(TSGetSNES(ts, &snes));
+          if (!A)
+            AssertTS(SNESSetUseMatrixFree(snes, PETSC_TRUE, PETSC_FALSE));
+
+          if (!P)
+            {
+              DM dm;
+              AssertTS(SNESGetDM(snes, &dm));
+              AssertTS(DMSetMatType(dm, MATSHELL));
+            }
+        }
+      else
+        {
+          if (explicit_jacobian)
+            {
+              AssertTS(TSSetRHSJacobian(ts,
+                                        A ? A->petsc_matrix() :
+                                            (P ? P->petsc_matrix() : nullptr),
+                                        P ? P->petsc_matrix() : nullptr,
+                                        _dealii_ts_rhsjacobian_,
+                                        nullptr));
+            }
+
+          if (implicit_jacobian)
+            {
+              AssertTS(TSSetIJacobian(ts,
+                                      A ? A->petsc_matrix() :
+                                          (P ? P->petsc_matrix() : nullptr),
+                                      P ? P->petsc_matrix() : nullptr,
+                                      _dealii_ts_ijacobian_,
+                                      nullptr));
+            }
+        }
+
+
+      // In case solve_for_jacobian_system is provided, create a shell
+      // preconditioner wrapping the user call. The internal Krylov
+      // solver will apply the preconditioner only once. This choice
+      // can be overriden by command line and users can use any other
+      // Krylov method if their solve is not accurate enough.
+      PreconditionShell precond(
+        PetscObjectComm(reinterpret_cast<PetscObject>(ts)));
+      if (solve_for_jacobian_system)
+        {
+          precond.vmult = [&](VectorBase &      indst,
+                              const VectorBase &insrc) -> int {
+            VectorType       dst(static_cast<const Vec &>(indst));
+            const VectorType src(static_cast<const Vec &>(insrc));
+            return solve_for_jacobian_system(src, dst);
+          };
+
+          SNES snes;
+          KSP  ksp;
+          AssertTS(TSGetSNES(ts, &snes));
+          AssertTS(SNESGetKSP(snes, &ksp));
+          AssertTS(KSPSetType(ksp, KSPPREONLY));
+          AssertTS(KSPSetPC(ksp, precond.get_pc()));
```

Comment:
I don't quite understand what this block does. If someone provides a solver for the Jacobian system, I would expect that that solver is actually used to solve the linear system. But I think what you are doing here is using this solver as a preconditioner for a KSP solver -- though perhaps for a special kind that only applies the preconditioner? Is this the PETSc way to fake a solve by creating a fake solver and putting the solver into the preconditioner? If so, would you mind expanding on this "trick" in the comment above?

Reply:
done